clippy_lints/src/mut_key.rs

```diff
@@ -103,30 +103,63 @@ fn check_sig<'tcx>(cx: &LateContext<'tcx>, item_hir_id: hir::HirId, decl: &hir::
 fn check_ty<'tcx>(cx: &LateContext<'tcx>, span: Span, ty: Ty<'tcx>) {
     let ty = ty.peel_refs();
     if let Adt(def, substs) = ty.kind() {
-        if [sym::hashmap_type, sym::BTreeMap, sym::hashset_type, sym::BTreeMap]
+        let is_keyed_type = [sym::hashmap_type, sym::BTreeMap, sym::hashset_type, sym::BTreeSet]
             .iter()
-            .any(|diag_item| cx.tcx.is_diagnostic_item(*diag_item, def.did))
-            && is_mutable_type(cx, substs.type_at(0), span)
-        {
+            .any(|diag_item| cx.tcx.is_diagnostic_item(*diag_item, def.did));
+        if is_keyed_type && is_mutable_type(cx, substs.type_at(0), span) {
             span_lint(cx, MUTABLE_KEY_TYPE, span, "mutable key type");
         }
     }
 }
 
 fn is_mutable_type<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>, span: Span) -> bool {
     match *ty.kind() {
-        RawPtr(TypeAndMut { ty: inner_ty, mutbl }) | Ref(_, inner_ty, mutbl) => {
-            mutbl == hir::Mutability::Mut || is_mutable_type(cx, inner_ty, span)
-        },
+        // Raw pointers are never considered 'mutable' by this lint, since there are two cases:
+        // 1. The raw pointer is hashed/ordered directly (e.g. in `HashSet<*const Cell<usize>>` or
+        //    `HashSet<*mut usize>`.  These impls depend only on the memory **address** of the
+        //    referred value, rather than the value itself.  Therefore, checking the contained type
+        //    causes a false positive.  Links (links refer to Rust version 1.54):
+        //      - clippy issue: https://github.com/rust-lang/rust-clippy/issues/6745
+        //      - hash impl: https://github.com/rust-lang/rust/blob/1.54.0/library/core/src/hash/mod.rs#L717-L736
+        //      - ord impls:
+        //        - https://github.com/rust-lang/rust/blob/master/library/core/src/ptr/const_ptr.rs#L1030-L1043
+        //        - https://github.com/rust-lang/rust/blob/master/library/core/src/ptr/mut_ptr.rs#L1345-L1357
+        //
+        // 2. The raw pointer is contained inside an ADT.  Using `#[derive(Hash)]` will call
+        //    `<*const T as Hash>::hash()`, which will have the same behaviour as case (1).  If the
+        //    ADT uses a custom `impl Hash` which (unsafely) dereferences the raw pointer, then
+        //    they are on their own - checking this in the general case would be impossible.
```

Comment:
This is not really relevant for this function since we are looking at the *type* and so we can't see ADT fields anyways. You could put this in a `Known problems` section of the lint docs.

Reply:
Done!